File: tasks/ember-tier-cap-audit/solution/quota_audit.exs
data = System.get_env("QUOTA_DATA_DIR", "/app/quota_lab")
audit = System.get_env("QUOTA_AUDIT_DIR", "/app/audit")
File.mkdir_p!(audit)

policy = data |> Path.join("policy.json") |> File.read!() |> Jason.decode!()
events = data |> Path.join("events.json") |> File.read!() |> Jason.decode!()
day = policy["audit_day"]
order = policy["tier_order"]

caps =
  policy["tier_caps"]
  |> Enum.map(fn {k, v} -> {k, trunc(v)} end)
  |> Map.new()

caps =
  Enum.reduce(events["tier_derates"] || [], caps, fn d, acc ->
    if d["start_day"] <= day and day <= d["end_day"] do
      t = d["tier"]
      if Map.has_key?(acc, t) do
        Map.put(acc, t, div(Map.fetch!(acc, t) * d["factor_bp"], 10000))
      else
        acc
      end
    else
      acc
    end
  end)

frozen =
  (events["item_freezes"] || [])
  |> Enum.filter(fn f -> f["start_day"] <= day and day <= f["end_day"] end)
  |> Enum.map(& &1["item_id"])
  |> MapSet.new()

tier_key = fn tier ->
  idx = Enum.find_index(order, &(&1 == tier))
  if idx, do: {idx, tier}, else: {length(order), tier}
end

items =
  data
  |> Path.join("items")
  |> File.ls!()
  |> Enum.filter(&String.ends_with?(&1, ".json"))
  |> Enum.sort()
  |> Enum.map(fn f ->
    [data, "items", f] |> Path.join() |> File.read!() |> Jason.decode!()
  end)
  |> Enum.sort_by(fn it -> Tuple.append(tier_key.(it["tier"]), it["item_id"]) end)

{tier_rem, rows, sc} =
  Enum.reduce(items, {caps, [], %{"frozen" => 0, "ok" => 0, "shortfall" => 0}}, fn it, {rem, acc, counts} ->
    iid = it["item_id"]
    tier = it["tier"]
    demand = trunc(it["demand"])

    if MapSet.member?(frozen, iid) do
      row = %{"item_id" => iid, "tier" => tier, "status" => "frozen", "demand" => demand, "allocated" => 0}
      {rem, acc ++ [row], Map.update!(counts, "frozen", &(&1 + 1))}
    else
      left = Map.get(rem, tier, 0)
      alloc = min(demand, left)
      rem2 = Map.put(rem, tier, left - alloc)
      st = if alloc == demand, do: "ok", else: "shortfall"
      row = %{"item_id" => iid, "tier" => tier, "status" => st, "demand" => demand, "allocated" => alloc}
      {rem2, acc ++ [row], Map.update!(counts, st, &(&1 + 1))}
    end
  end)

touched =
  rows
  |> Enum.filter(fn r -> r["allocated"] > 0 end)
  |> Enum.map(& &1["tier"])
  |> Enum.uniq()
  |> Enum.sort()

summary = %{
  "audit_day" => day,
  "items_processed" => length(items),
  "frozen_items" => sc["frozen"],
  "status_counts" => sc,
  "tiers_touched" => touched
}

encode = fn obj ->
  Jason.encode!(obj, pretty: true) <> "\n"
end

File.write!(Path.join(audit, "allocations.json"), encode.(%{"items" => rows}))
File.write!(Path.join(audit, "summary.json"), encode.(summary))

File: tasks/hazel-tier-cap-audit/solution/quota_audit.mjs
import fs from "node:fs";
import path from "node:path";

const data = process.env.QUOTA_DATA_DIR ?? "/app/quota_lab";
const audit = process.env.QUOTA_AUDIT_DIR ?? "/app/audit";
fs.mkdirSync(audit, { recursive: true });

const policy = JSON.parse(fs.readFileSync(path.join(data, "policy.json"), "utf8"));
const events = JSON.parse(fs.readFileSync(path.join(data, "events.json"), "utf8"));
const day = policy.audit_day;
const order = policy.tier_order;

const caps = Object.fromEntries(
  Object.entries(policy.tier_caps).map(([k, v]) => [k, Number(v)])
);

for (const d of events.tier_derates ?? []) {
  if (d.start_day <= day && day <= d.end_day && caps[d.tier] !== undefined) {
    caps[d.tier] = Math.floor((caps[d.tier] * d.factor_bp) / 10000);
  }
}

const frozen = new Set(
  (events.item_freezes ?? [])
    .filter((f) => f.start_day <= day && day <= f.end_day)
    .map((f) => f.item_id)
);

const tierKey = (tier) => {
  const idx = order.indexOf(tier);
  return idx >= 0 ? [idx, tier] : [order.length, tier];
};

const items = fs
  .readdirSync(path.join(data, "items"))
  .filter((f) => f.endsWith(".json"))
  .sort()
  .map((f) => JSON.parse(fs.readFileSync(path.join(data, "items", f), "utf8")))
  .sort((a, b) => {
    const [ra, ta] = tierKey(a.tier);
    const [rb, tb] = tierKey(b.tier);
    if (ra !== rb) return ra - rb;
    if (ta !== tb) return ta.localeCompare(tb);
    return a.item_id.localeCompare(b.item_id);
  });

const tierRem = { ...caps };
const rows = [];
const sc = { frozen: 0, ok: 0, shortfall: 0 };

for (const it of items) {
  const demand = Number(it.demand);
  if (frozen.has(it.item_id)) {
    rows.push({
      item_id: it.item_id,
      tier: it.tier,
      status: "frozen",
      demand,
      allocated: 0,
    });
    sc.frozen += 1;
    continue;
  }
  const left = tierRem[it.tier] ?? 0;
  const alloc = Math.min(demand, left);
  tierRem[it.tier] = left - alloc;
  const st = alloc === demand ? "ok" : "shortfall";
  sc[st] += 1;
  rows.push({
    item_id: it.item_id,
    tier: it.tier,
    status: st,
    demand,
    allocated: alloc,
  });
}

const touched = [...new Set(rows.filter((r) => r.allocated > 0).map((r) => r.tier))].sort();
const summary = {
  audit_day: day,
  items_processed: items.length,
  frozen_items: sc.frozen,
  status_counts: sc,
  tiers_touched: touched,
};

const write = (name, obj) => {
  fs.writeFileSync(path.join(audit, name), `${JSON.stringify(obj, null, 2)}\n`, "utf8");
};

write("allocations.json", { items: rows });
write("summary.json", summary);
